src/Aspire.Hosting/ApplicationModel/ExpressionResolver.cs

```diff
@@ -77,24 +77,41 @@ await EvalEndpointAsync(endpointReference, EndpointProperty.Port).ConfigureAwait
         };
     }
 
-    async Task<string?> EvalExpressionAsync(ReferenceExpression expr)
+    async Task<ResolvedValue> EvalExpressionAsync(ReferenceExpression expr)
     {
         // This logic is similar to ReferenceExpression.GetValueAsync, except that we recurse on
         // our own resolver method
         var args = new object?[expr.ValueProviders.Count];
+        var isSensitive = false;
 
         for (var i = 0; i < expr.ValueProviders.Count; i++)
         {
-            args[i] = await ResolveInternalAsync(expr.ValueProviders[i]).ConfigureAwait(false);
+            var result = await ResolveInternalAsync(expr.ValueProviders[i]).ConfigureAwait(false);
+            args[i] = result?.Value;
+            if (result?.IsSensitive is true)
+            {
+                isSensitive = true;
+            }
         }
 
-        return string.Format(CultureInfo.InvariantCulture, expr.Format, args);
+        return new(string.Format(CultureInfo.InvariantCulture, expr.Format, args), isSensitive);
     }
 
-    async Task<string?> EvalValueProvider(IValueProvider vp)
+    async Task<ResolvedValue> EvalValueProvider(IValueProvider vp)
     {
         var value = await vp.GetValueAsync(cancellationToken).ConfigureAwait(false);
 
+        if (vp is ParameterResource pr)
+        {
+            return new ResolvedValue(value, pr.Secret);
+        }
+
+        // No need to do extra work, since the below will only be valid for containers.
+        if (!sourceIsContainer)
```

Comment:
The `sourceIsContainer` parameter complicates the `ResolveInternalAsync` logic below. Could this parameter be eliminated? What do we gain by having it? For example, is value provider (`vp`) going to ever be `HostUrl` if we are not looking at a container?

Reply:
It seems like yes, `WithOtlpExporter`  -> `AddOtlpEnvironment` will add a HostUrl, which can be resolved even in executables. 

The goal of the parameter was to avoid the value substitution on executables, to preserve the previous behavior